Modules/Module_Import/NII2BIDS/xASL_bids_PhoenixProtocolAnalyzer.m

```diff
@@ -14,198 +14,256 @@
+			bidsPar.SoftwareVersions = strrep(sourcePar.sProtConsistencyInfotBaselineString,'"','');
 		end
-	end
-	% The function of this parameter is unclear
-	%if isfield(sourcePar,'Slab_thickness_mm') && ~isempty(sourcePar.Slab_thickness_mm)
-	%	bidsPar.LabelingSlabThickness = sourcePar.Slab_thickness_mm;
-	%end
-	
-	if isfield(sourcePar,'alTE0') && ~isempty(sourcePar.alTE0)
-		bidsPar.EchoTime = sourcePar.alTE0 / 1000 / 1000;
-		lastEchoTime = bidsPar.EchoTime;
-		lastEchoTimeCount = 1;
-		while lastEchoTime > 0
-			lastEchoTimeCount = lastEchoTimeCount + 1;
-			echoTimePar = getPhoenixParameters({['alTE[' num2str(lastEchoTimeCount) ']']},parameterList,0);
-			if isempty(echoTimePar{2})
-				lastEchoTime = 0;
-			else
-				lastEchoTime = str2num(echoTimePar{2})/1000/1000;
-				bidsPar.EchoTime(lastEchoTimeCount) = lastEchoTime;
+
+		% N4_VD13D and N4_VE11C have also slab_thickness_mm parameter defined and it is unclear if this is imaging FOV or labeling slab thickness...
+		if isfield(bidsPar,'SoftwareVersions') &&...
+				(~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VB15A', 'once'))||...
+				~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VB17A', 'once'))||...
+				~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VB19A', 'once')))
+			if isfield(sourcePar,'sGroupArraysPSatdGap') && ~isempty(sourcePar.sGroupArraysPSatdGap)
+				bidsPar.LabelingDistance = sourcePar.sGroupArraysPSatdGap;
 			end
-		end		
-		
-	end
-	
-	% If the labeling type is recognized, then proceed to labeling timing information extraction
-	if isfield(bidsPar,'ArterialSpinLabelingType') 
-		if~isempty(regexpi(bidsPar.ArterialSpinLabelingType,'pasl'))
-			if isfield(bidsPar,'SoftwareVersions') &&...
-					(~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VB15A'))||...
-					 ~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VB17A'))||...
-					 ~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VB19A'))||...
-					 ~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VD13D'))||...
-					 ~isempty(regexpi(bidsPar.SoftwareVersions,'N4_VE11C')))
-				
-				if isfield(sourcePar,'alTI0') && ~isempty(sourcePar.alTI0)
-					bidsPar.PostLabelingDelay = sourcePar.alTI0 / 1000 / 1000;
+
+			if isfield(bidsPar,'ArterialSpinLabelingType') && ~isempty(regexpi(bidsPar.ArterialSpinLabelingType,'pasl', 'once')) &&...
+					isfield(sourcePar,'sGroupArraysPSatdThickness') && ~isempty(sourcePar.sGroupArraysPSatdThickness)
+				bidsPar.LabelingSlabThickness = sourcePar.sGroupArraysPSatdThickness;
+			end
+		end
+		% The function of this parameter is unclear
+		%if isfield(sourcePar,'Slab_thickness_mm') && ~isempty(sourcePar.Slab_thickness_mm)
+		%	bidsPar.LabelingSlabThickness = sourcePar.Slab_thickness_mm;
+		%end
+
+		if isfield(sourcePar,'alTE0') && ~isempty(sourcePar.alTE0)
+			bidsPar.EchoTime = sourcePar.alTE0 / 1000 / 1000;
+			lastEchoTime = bidsPar.EchoTime;
+			lastEchoTimeCount = 1;
+			while lastEchoTime > 0
```

Comment:
no risk ending up in an infinite loop?

echoTimePar{2} is always empty in the last case, cannot be an empty struct/cell array that matlab sees as non-empty?

Reply:
Exactly - no risk. This works and was tested.
It reads alTE[XX] from Phoenix. It ends when there are zeros or when the field ends. Absolutely no risk this ends in an infinite loop.